File: tools/enums.rb
#!/usr/bin/env ruby 
require "rtruckboris"

LICENSE = <<EOF
/*
 * lua-clangc lua bindings for the C interface of Clang
 * Copyright (C) 2015  cedlemo <cedlemo@gmx.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
EOF

clang_c = "/usr/include/clang-c/Index.h"
headerPaths = []
#get gcc include if needed
gccLibBase='/usr/lib/gcc/' << `llvm-config --host-target`.chomp << "/*"
gccLib = Dir.glob(gccLibBase ).sort.last + "/include"
headerPaths << gccLib
headerPaths << "/usr/include"
headerPaths << "/usr/include/clang-c"

parser = Rtruckboris::HeaderParser.new(clang_c, headerPaths)
if(!parser.parse(true))
  puts "Whada fock??"
  exit
end
def create_table_constant(enum)
  name = enum.name == "" ? enum.typedef_name : enum.name
  enum_module_name = name.gsub(/CX/,"")
  s="lua_newtable(L);\n"
  enum.enum_constants.each do |c|
    s+="lua_pushinteger(L, #{c.value});\n"
    s+="lua_setfield(L, -2, \"#{c.name.gsub(/CX[^\_]*\_/,"").capitalize}\");\n"
  end 
  s+="lua_setfield(L, -2,\"#{enum_module_name}\");\n"
  s
end
def create_tables_constants(enums)
  s="#include \"constants.h\"\n" 
  s+="void init_clang_enums_to_constants(lua_State *L){\n"
  enums.each_with_index do |e, i|
    s+="#{create_table_constant(e)}\n"
  end
  s+="}\n"
end
enums = parser.enums

class OutputFiles
  attr_reader :_c, :_h
  def initialize(name)
    @_c = File.new((name||'wrapper') + '.c', 'w')
    @_h = File.new((name||'wrapper') + '.h', 'w')
    @_c.sync= true
    @_c.sync= true
  end
  def close_all
    @_c.close
    @_h.close
  end
end 
constants = OutputFiles.new("constants")
constants._h.puts LICENSE
constants._c.puts LICENSE

constants._h.puts <<EOF
#include "lua.h" 
void init_clang_enums_to_constants(lua_State *L);
EOF

clang_c_errors = "/usr/include/clang-c/CXErrorCode.h"
parser = Rtruckboris::HeaderParser.new(clang_c_errors, headerPaths)
parser.parse(true)

enums += parser.enums

constants._c.puts create_tables_constants(enums)
constants.close_all

File: tools/enums.lua
#!/usr/bin/env lua

local toolbox = require('toolbox')

local gcc_lib_base = '/usr/lib/gcc/' .. toolbox.exec_out('llvm-config --host-target')

local last = ''
for dir in toolbox.lfs.dir(gcc_lib_base) do
  if not dir:match('^..?$') then
    if dir > last then
      last = dir
    end
  end
end

headers_paths = {}
headers_paths[1] = gcc_lib_base .. '/' .. last .. '/' .. 'include'
headers_paths[2] = '/usr/include'
headers_paths[3] = '/usr/include/clang-c'

clangc_source = '/usr/include/clang-c/Index.h'

local parser = toolbox.truckboris.HeaderParser()
parser:add_source_file(clangc_source)
for _,h in ipairs(headers_paths) do
  parser:add_search_path(h)
end

has_worked = parser:parse()

if has_worked == true then
  print("Parsing has worked")
end

local enums = {}

for i = 1, parser:enums_num(), 1 do
  enums[i] = parser:get_nth_enum(i - 1)
end

local function rename_module_table(enum)
  local name = enum:name() == "" and enum:typedef_name() or enum:name()
  return name:gsub('^CX_?','') 
end

local function rename_enum_constant(constant_name)
  local name = constant_name:gsub("^([^_]+_)(.*)$","%2")
  return toolbox.to_screaming_snake_case(name)
end

local function create_module_constant(enum)
  local name = rename_module_table(enum)
  print('-->' .. name)
  for i = 1, enum:constants_num(), 1 do
    local enum_constant_name = enum:get_nth_enum_constant(i - 1):name()
    print('\t-' .. rename_enum_constant(enum_constant_name) .. ' -- ' .. enum_constant_name)
  end
end

for _,e in ipairs(enums) do
  create_module_constant(e)
end
